File: algs/sort/sort.js
function swap(array, i, j) {
    const tmp = array[i];
    array[i] = array[j];
    array[j] = tmp;
}

// next line 없이 콘솔 출력. 귀찮아서 작성
const print = (obj) => { process.stdout.write(obj.toString()) }
// 배열을 줄바꿈 없이 출력(console.log 찍으면 이쁘게 정렬되는걸 막음)
const printArr = (array) => {
    const sz = array.length;
    print("{ ");
    if (sz >= 1)  print(array[0]);
    for (let i = 1; i < sz; ++i) {
        print(", ");
        print(array[i]);
    }
    print(" }\n");
}

// sort: 함수 선언식, 함수 표현식, 화살표함수등을 골고루 섞어서 선언하였으며, 반복문은 while, for 문을 바꿔서 쓰기도 함.

function bubble(array) {
    const size = array.length;
    for (let i = 0; i < size - 1; ++i) {
        for (let j = 0; j < size - i - 1; ++j) {
            if (array[j] > array[j + 1])        swap(array, j, j + 1);
        }
    }
}

const select = function(array) {
    const size = array.length;
    for (let i = 0; i < size; i++) {
        let minIdx = i;
        let j = i;
        while (j < size) {
            if (array[j] < array[minIdx])       minIdx = j;
            ++j;
        }
        if (minIdx !== i)                        swap(array, i, minIdx);
    }
}

const insert = (array) => {
    const size = array.length;
    for (let i = 1; i < size; ++i) {
        let bf = array[i];
        let j = i; 
        while (j > 0 && array[j - 1] > bf)      array[j] = array[--j];
        if (j != i)                             array[j] = bf;
    }
}

const shell = function(array) {
    const size = array.length;
    let step = 1;
    while (step < Math.floor(size / 3))
        step = step * 3 + 1;

    for (;step > 0; step = Math.floor(step /= 3)) {
        for (let i = step; i < size; ++i) {
            const bf = array[i];
            let j;
            for (j = i; j >= step && array[j - step] > bf; j -= step)
                array[j] = array[j - step];
            if (j !== i)
                array[j] = bf;
        }
    
    }
}


// TODO quick 전용 전환 삽입정렬. 인레이 구현 등 리펙토링 필요
function insertSort(array, startIdx, endIdx) {
    for (let i = startIdx + 1; i <= endIdx; i++) {
        let key = array[i];
        let j = i - 1;
        while (j >= startIdx && array[j] > key) {
            array[j + 1] = array[j];
            j--;
        }
        array[j + 1] = key;
    }
}

const _quick = function(array, startIdx, endIdx) {
    if (endIdx - startIdx + 1 <= 256) {
        insertSort(array, startIdx, endIdx);
        return;
    }

    let leftIdx = startIdx;
    let rightIdx = endIdx;
    const centerIdx = startIdx + Math.floor((endIdx - startIdx) / 2);
    const pivot = array[centerIdx];
    
    while (leftIdx <= rightIdx) {
        while (array[leftIdx] < pivot)      leftIdx++;
        while (array[rightIdx] > pivot)     rightIdx--;
        if (leftIdx <= rightIdx)            swap(array, leftIdx++, rightIdx--);
    }

    if (startIdx < rightIdx)                _quick(array, startIdx, rightIdx);
    if (endIdx > leftIdx)                   _quick(array, leftIdx, endIdx);
}

const quick = function(array) { _quick(array, 0, array.length - 1) }

const merge = (array) => {
    const arrSz = array.length;
    if (arrSz <= 1) return;
    
    if (arrSz <= 256) {
        insert(array);
        return;
    }

    // divide
    const leftSz = Math.floor(arrSz / 2);
    const rightSz = arrSz - leftSz;
    const leftArr = array.slice(0, leftSz)
    const rightArr = array.slice(leftSz)

    // conquer
    merge(leftArr);
    merge(rightArr);

    // merge
    // 좌우측 배열에서 순서대로 뽑아 array에 재정렬
    let lIdx = 0,  rIdx = 0,  aIdx = 0;

    while (lIdx < leftSz && rIdx < rightSz) {
        if (leftArr[lIdx] < rightArr[rIdx])     array[aIdx++] = leftArr[lIdx++]
        else                                    array[aIdx++] = rightArr[rIdx++]
    }

    // 나머지 채워넣기
    while (lIdx < leftSz)       array[aIdx++] = leftArr[lIdx++]
    while (rIdx < rightSz)      array[aIdx++] = rightArr[rIdx++]
}

const getRandInt = (max) => {
    return Math.floor(Math.random() * max) + 1
}

const getRandIntArray = (size, max = 10000) => {
    res = []
    for (let i = 0; i < size; ++i) {
        res.push(getRandInt(max))
    }
    return res
}

const main = () => {

    // 출력
    {
        let sample = [ 2, 3, 7, 1, 9, 6, 0, 5, 4, 8 ];
        printArr(sample);
        insert(sample)
        printArr(sample);
    }

    // 두 정렬 비교출력(길이 50 미만시)
    {
        const sampleSize = 100;
        let randSample = getRandIntArray(sampleSize, 300);
        let cp = randSample.slice()
        if (sampleSize <= 50) printArr(randSample);

        quick(randSample);
        // shell(randSample);
        cp.sort((a, b) => a - b);
        

        console.log((
            () => {
            for (let i = 0; i < sampleSize; ++i) {
                if (cp[i] != randSample[i]) {
                    if (sampleSize <= 50) {
                        printArr(randSample);
                        printArr(cp);
                    }
                    else {
                        console.log(`${i}번 index ${cp[i]} != ${randSample[i]}`);
                    }
                    return "[FAIL] 정렬 결과 같지않음";
                }
            }
            return "[SUCCESS] 정렬 결과 같음";
        })());
    }
    
    // 단일시간측정
    {
        const func = merge;
        const sampleSize = 100000000;

        let randSample = getRandIntArray(sampleSize, 10000)
        let startTime = performance.now()
        func(randSample)
        // randSample.sort((a, b) => a - b);       // 기본정렬 사용시
        const resTime = performance.now() - startTime
        print(resTime / 1000 + " s\n")
    }
}

main()

File: algs/sort/Sort.java
import java.util.Random;
import java.util.function.BiConsumer;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.Map;

// ArrayList 사용시 배열 접근이 너무 느려서 전통적 배열로 작성.
// 퀵정렬 / 병합정렬에서 insert를 호출시키기 위해 index기반 정렬도 구현됨
// 병합정렬은 기초구현(좌우측 배열 재할당) 방식과 sbMerge(버퍼를 미리 한번 할당하여 재활용)방식이 구현됨.


class Sort {
    // ---- HELPER: 가독성을 위해 함수를 모두 축소시키는 것을 추천.  ----

    public static void print(int[] arr) {
        System.out.print("[ ");
        int sz =                        arr.length;
        if (sz > 0)                     System.out.print(arr[0]);
        for (int i = 1; i < sz; ++i)    System.out.print(", " + arr[i]);
        System.out.print(" ]\n");
    }

    public static void swap(int[] arr, int i, int j) {
        int buffer =    arr[i];
        arr[i] =        arr[j];
        arr[j] =        buffer;
    }

    public static boolean isNeedSwap(int a, int b, boolean ascending) {
        return ascending? a > b : a < b;
    }

    public static int[] getRandArr(int size, int max) {
        Random random =                 new Random();
        int[] res =                     new int[size];
        for (int i = 0; i < size; ++i)  res[i] = random.nextInt(max) + 1;
        return res;
    }

    public static void doTest
    (
        // parameter
        int[]                       sample,
        BiConsumer<int[], Boolean>  algorithm,
        String                      funcName
    ) 
    {   
        // definition
        boolean ascending =             false;
        int[] copy =                    sample.clone();
        long start =                    System.currentTimeMillis();
        algorithm.accept(copy, ascending);
        long spent =                    System.currentTimeMillis() - start;

        System.out.println(
            "sort: " + funcName + "\t sample size: " + (copy.length) + "\tspent: " + spent + " ms \t (" + (spent / 1000.0) + " s)"
        );
    }

    public static boolean isCorrectlySorted(int[] sorted, int[] originalSample, boolean ascending) {
        // 정렬검사: originalSample 원본 정렬함.
        javaSort(originalSample, ascending);
        return Arrays.equals(sorted, originalSample);
    }


    // ---- SORT ALGORITHM ----

    public static void bubble(int[] arr, boolean ascending) {
        int sz = arr.length;
        for (int i = 0; i < sz - 1; ++i) {
            for (int j = 0; j < sz - i - 1; ++j) {
                if (isNeedSwap(arr[j], arr[j + 1], ascending))
                    swap(arr, j, j + 1);
            }
        }
    }

    public static void select(int[] arr, boolean ascending) {
        int sz = arr.length;
        for (int i = 0; i < sz; ++i) {
            int idxBuffer = i;
            for (int j = i; j < sz; ++j) {
                if (isNeedSwap(arr[idxBuffer], arr[j], ascending))
                    idxBuffer = j;
            }
            if (idxBuffer != i)
                swap(arr, i, idxBuffer);
        }
    }

    public static void insert(int[] arr, boolean ascending) {
        int sz = arr.length;
        for (int i = 1; i < sz; ++i) {
            int buffer = arr[i];
            int j;
            for (j = i; j > 0 && isNeedSwap(arr[j - 1], buffer, ascending); --j)
                arr[j] = arr[j - 1];
            if (j != i)
                arr[j] = buffer;
        }
    }

    public static void shell(int[] arr, boolean ascending) {
        int sz = arr.length;
        int step = 1;
        while (step < sz / 3) step = step * 3 + 1;

        while (step > 0) {
            for (int i = step; i < sz; ++i) {
                int buffer = arr[i];
                int j;
                for (j = i; j >= step && isNeedSwap(arr[j - step], buffer, ascending); j -= step)
                    arr[j] = arr[j - step];
                if (j != i)
                    arr[j] = buffer;
            }
            step /= 3;
        }
    }

    public static void _insert(int[] arr, int startIdx, int endIdx, boolean ascending) {
        // 삽입정렬:index기반 - quick,  merge 등 에서 쉽게 사용하기 위함
        for (int i = startIdx + 1; i <= endIdx; ++i) {
            int bf = arr[i];
            int j;
            for (j = i; j > startIdx && isNeedSwap(arr[j - 1], bf, ascending); --j)
                arr[j] = arr[j - 1];
            if (j != i)
                arr[j] = bf;
        }
    }

    public static void _quick(int[] arr, int startIdx, int endIdx, boolean ascending) {
        // 아주 작은 배열이 재귀 실행되는것을 막기 위해 크기가 작아지면 insert로 전환
        if (endIdx - startIdx <= 256) { _insert(arr, startIdx, endIdx, ascending);      return;}

        int lIdx = startIdx;
        int rIdx = endIdx;
        int pivot = arr[(lIdx + rIdx) / 2];

        while (lIdx <= rIdx) {
            while (ascending ?
                    arr[lIdx] < pivot :
                    arr[lIdx] > pivot) 
                    lIdx++;
            while (ascending ?
                    arr[rIdx] > pivot :
                    arr[rIdx] < pivot) 
                    rIdx--;
            if (lIdx <= rIdx) {
                swap(arr, lIdx, rIdx);
                lIdx++;
                if (rIdx > 0) rIdx--;       // idx 가 unsigned 자료형이라면 필요한 조건식임. 정수형은 rIdx--만으로도 충분
            }
        }

        if (startIdx < rIdx) _quick(arr, startIdx, rIdx, ascending);
        if (endIdx > lIdx) _quick(arr, lIdx, endIdx, ascending);
    }

    public static void quick(int[] arr, boolean ascending) {
        _quick(arr, 0, arr.length - 1, ascending);
    }

    public static void _merge(int[] arr, int[] left, int[] right, boolean ascending) {
        int aSize = arr.length, lSize = left.length, rSize = right.length;
        int aIdx = 0, lIdx = 0, rIdx = 0;

        while (aIdx < aSize && lIdx < lSize && rIdx < rSize) {
            boolean cond =      left[lIdx] < right[rIdx];
                    cond =      ascending ? cond : !cond;

            if (cond)           arr[aIdx++] =   left[lIdx++];
            else                arr[aIdx++] =   right[rIdx++];
        }

        while (lIdx < lSize)    arr[aIdx++] = left[lIdx++];
        while (rIdx < rSize)    arr[aIdx++] = right[rIdx++];
    }

    public static void merge(int[] arr, boolean ascending) {
        int sz = arr.length;
        if (sz <= 1) return;
        
        if (sz <= 128) {
            insert(arr, ascending);
            return;
        }

        int lSize = sz / 2;

        int[] left = new int[lSize];
        int[] right = new int[sz - lSize];
        for (int i = 0; i < sz; ++i) {
            if (i < lSize) left[i] = arr[i];
            else right[i - lSize] = arr[i];
        }

        merge(left, ascending);
        merge(right, ascending);

        _merge(arr, left, right, ascending);
    }


    public static void _sbMerge(int[] arr, int startIdx, int endIdx, int[] bf, boolean ascending) {
        // 기저조건
        int sz = endIdx - startIdx + 1;
        if (sz <= 1) return;
        if (sz <= 64) {
            _insert(arr, startIdx, endIdx, ascending);
            return;
        }

        // 분할위치선정
        int leftSz = (int)(sz / 2);
        int rightSz = sz - leftSz;

        // 정복
        _sbMerge(arr, startIdx, startIdx + leftSz - 1, bf, ascending);
        _sbMerge(arr, startIdx + leftSz, startIdx + sz - 1, bf, ascending);

        // 합병
        int lBegin = startIdx, rBegin = startIdx + leftSz, bIdx = startIdx;         // 시작점 인덱스기반,
        int lEnd = startIdx + leftSz, rEnd = lEnd + rightSz, bEnd = endIdx;         // 주의: 도달하면 안됨
        while (lBegin < lEnd && rBegin < rEnd && bIdx < bEnd) {
            boolean cond = ascending? arr[lBegin] < arr[rBegin] : arr[lBegin] > arr[rBegin];
            if (cond) {
                bf[bIdx++] = arr[lBegin++];
            } else {
                bf[bIdx++] = arr[rBegin++];
            }
        }

        while (lBegin < lEnd)   bf[bIdx++] = arr[lBegin++];
        while (rBegin < rEnd)   bf[bIdx++] = arr[rBegin++];

        // 결과 적용 bf -> arr
        for (int i = startIdx; i <= endIdx; ++i)    arr[i] = bf[i];
        // System.arraycopy(bf, startIdx, arr, startIdx, endIdx - startIdx + 1);    // 아주 미세한 성능향상

    }


    public static void sbMerge(int[] arr, boolean ascending) {
        // 버퍼공간을 미리 할당하여 재사용.
        int sz = arr.length;
        int[] bf = new int[sz];
        _sbMerge(arr, 0, sz - 1, bf, ascending);
    }

    // 속도비교를 위한 기본 Arrays의 정렬
    public static void javaSort(int[] arr, boolean ascending) {
        java.util.Arrays.sort(arr);

        // 억지로 구현한 내림차순. 작동은 되나 가능하면 리펙토링 필요
        if (!ascending) {
            int arrSize = arr.length;
            int[] cp = new int[arrSize];
            for (int i = 0; i < arrSize; ++i) {
                cp[i] = arr[arrSize - i - 1];
            }
            
            for (int i = 0; i < arrSize; ++i) {
                arr[i] = cp[i];
            }

        }
    }

    //! MAIN

    public static void main(String[] args) {
        
        boolean[] test = { 
            // 테스트 실행 스위치. 필요한 것만 최소로 골라 실행할 수록 측정결과가 좋게 나오는 경향이 있음(특히 sbMerge)
            true, true, true, true 
        };

        // 1. 출력 육안확인
        if (test[0]) { 
            System.out.println("\n--- 1. print sorted ---");
            int[] sample = getRandArr(10, 100);
            print(sample);
            sbMerge(sample, true);
            print(sample);
            sbMerge(sample, false);
            print(sample);
        }

        // 2. 정렬 신뢰성 검증 (비교정렬은 isCorrectlySorted 내부에서 java.util.Arrays.sort() 를 사용함)
        if (test[1]) {
            System.out.println("\n--- 2. validate the sort ---");
            BiConsumer<int[], Boolean> sortAlg = Sort::     sbMerge;      // 변경가능

            int sampleSize = 30000;
            int[] sample = getRandArr(sampleSize, 10000);
            int[] cpSample = sample.clone();

            sortAlg.accept(sample, true);                                           // 정렬진행
            boolean res = isCorrectlySorted(sample, cpSample, true);     // 비교정렬 진행후 결과산출
            System.out.println(String.format("[오름차순 %s]: 길이 %d 배열", res?"성공":"실패", sampleSize));
            
            if (res) {  // 오름차순이 성공하면 진행됨. 샘플 오염된 채로 그대로 다시 사용(개선할 필요 없음. 어차피 재정렬)
                sortAlg.accept(sample, false);
                res = isCorrectlySorted(sample, cpSample, false);
                System.out.println(String.format("[내림차순 %s]: 길이 %d 배열", res?"성공":"실패", sampleSize));
            }
        }

        // 3. 기본 벤치마크
        if (test[2]) {
            System.out.println("\n--- 3. benchmark sorting ---");
            int[] sample = getRandArr(30000, 10000);
            Map<String, BiConsumer<int[], Boolean>> sortAlgs = new LinkedHashMap<>();
            sortAlgs.put("bubble", Sort::bubble);
            sortAlgs.put("select", Sort::select);
            sortAlgs.put("insert", Sort::insert);
            sortAlgs.put("shell", Sort::shell);
            sortAlgs.put("merge", Sort::merge);
            sortAlgs.put("quick", Sort::quick);
            sortAlgs.put("sbMerge", Sort::sbMerge);

            for (String key : sortAlgs.keySet()) {
                doTest(sample, sortAlgs.get(key), key);
            }
        }

        // 4. 고부하 정렬 벤치마크
        if (test[3]) {
            System.out.println("\n--- 4. High Perfomance Sort ---");
            int[] sample = getRandArr(100000000, 10000);
            doTest(sample, Sort::merge, "merge");
            doTest(sample, Sort::sbMerge, "sbMerge");
            doTest(sample, Sort::quick, "quick");
            doTest(sample, Sort::javaSort, "javaSort");
        }
    }
}

